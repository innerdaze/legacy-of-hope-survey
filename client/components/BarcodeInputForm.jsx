import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Form from 'grommet/components/Form'
import Button from 'grommet/components/Button'
import Box from 'grommet/components/Box'
import Modes from '../constants/OperationModes'
import { orderSelectors } from '../features/order'
class BarcodeInputForm extends Component {
  constructor(props) {
    super(props)

    this.state = {
      barcode: ''
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
    this.receiveFocus = this.receiveFocus.bind(this)
    this.assignInputRef = this.assignInputRef.bind(this)
  }

  componentDidMount() {
    this.receiveFocus()
  }

  componentWillReceiveProps(newProps) {
    if (
      this.props.shouldFocusField !== newProps.shouldFocusField &&
      newProps.shouldFocusField
    ) {
      this.receiveFocus()
    }
  }

  handleChange(e) {
    this.setState({
      barcode: e.target.value
    })
  }

  receiveFocus() {
    setTimeout(() => {
      this.inputRef.focus()
      this.inputRef.select()
    }, 100)
  }

  handleSubmit(e) {
    e.preventDefault()
    this.props.onSubmitBarcode(this.state.barcode)
    this.setState({
      barcode: ''
    })
  }

  assignInputRef(ref) {
    this.inputRef = ref
  }

  render() {
    const css = `
      #barcodeInputContainer .grommetux-button__icon {
        display: none;
      }

      #barcodeInputContainer .grommetux-number-input {
        display: block;
      }

      #barcodeInputContainer input {
        width: 100%;
      }

      input[type="number"]::-webkit-outer-spin-button,
      input[type="number"]::-webkit-inner-spin-button {
          -webkit-appearance: none;
          margin: 0;
      }
      input[type="number"] {
          -moz-appearance: textfield;
      }
    `
    return (
      <Form plain>
        <style>{css}</style>
        <Box direction='row' alignContent='stretch'>
          <div id='barcodeInputContainer'>
            <input
              autoFocus
              placeholder='Enter barcode'
              value={this.state.barcode}
              ref={this.assignInputRef}
              onChange={this.handleChange}
            />
          </div>
          <Button
            secondary
            fill
            label='Search'
            type='submit'
            onClick={this.handleSubmit}
          />
        </Box>
      </Form>
    )
  }
}

BarcodeInputForm.propTypes = {
  onSubmitBarcode: PropTypes.func.isRequired,
  shouldFocusField: PropTypes.bool.isRequired
}

BarcodeInputForm.defaultProps = {
  onSubmitBarcode: Function.prototype,
  shouldFocusField: PropTypes.bool.isRequired
}

export default BarcodeInputForm
