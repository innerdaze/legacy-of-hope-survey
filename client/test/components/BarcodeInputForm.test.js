import React from 'react'
import BarcodeInputForm from '../../components/BarcodeInputForm'

let wrapper, onSubmitBarcode, shouldFocusField

jest.useFakeTimers()

describe('Testing on BarcodeInputForm...', () => {
  beforeEach(() => {
    onSubmitBarcode = jest.fn()
    shouldFocusField = false

    wrapper = shallow(
      <BarcodeInputForm
        onSubmitBarcode={onSubmitBarcode}
        shouldFocusField={shouldFocusField}
      />
    )
  })

  describe('BarcodeInputForm layout', () => {
    test('Expect BarcodeInputForm to have a Form with an input field and Button as child comps', () => {
      expect(wrapper.find('Form')).toHaveLength(1)
      expect(wrapper.find('Form').find('input')).toHaveLength(1)
      expect(wrapper.find('Form').find('Button')).toHaveLength(1)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Test on prop', () => {
    test('Expect onSubmitBarcode prop function is called when submit Button is clicked', () => {
      wrapper.find('Button').simulate('click', {
        preventDefault() {}
      })

      expect(onSubmitBarcode).toBeCalled()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Test on state', () => {
    test('Expect barcode state to change when input field is changed', () => {
      wrapper.find('input').prop('onChange', {
        preventDefault() {}
      })({
        target: {
          value: 'foo'
        }
      })

      expect(wrapper.state('barcode')).toEqual('foo')
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Test on methods', () => {
    test('Expect componentDidMount and receiveFocus is called', () => {
      let componentDidMountSpy = jest.spyOn(
        BarcodeInputForm.prototype,
        'componentDidMount'
      )
      let receiveFocusSpy = jest.spyOn(
        BarcodeInputForm.prototype,
        'receiveFocus'
      )
      wrapper = mount(
        <BarcodeInputForm
          onSubmitBarcode={onSubmitBarcode}
          shouldFocusField={shouldFocusField}
        />
      )

      expect(componentDidMountSpy).toHaveBeenCalled()
      expect(receiveFocusSpy).toHaveBeenCalled()
      expect(setTimeout.mock.calls.length).toBeGreaterThanOrEqual(1)
      expect(wrapper).toMatchSnapshot()
    })

    test('Expect componentWillReceiveProps is called when shouldFocusField change from false to true', () => {
      let componentWillReceivePropsSpy = jest.spyOn(
        BarcodeInputForm.prototype,
        'componentWillReceiveProps'
      )
      wrapper = mount(
        <BarcodeInputForm
          onSubmitBarcode={onSubmitBarcode}
          shouldFocusField={shouldFocusField}
        />
      )

      wrapper.setProps({
        shouldFocusField: true
      })

      expect(componentWillReceivePropsSpy).toHaveBeenCalled()
      expect(wrapper).toMatchSnapshot()
    })
  })
})
