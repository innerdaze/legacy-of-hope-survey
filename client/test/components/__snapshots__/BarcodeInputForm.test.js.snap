// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on BarcodeInputForm... BarcodeInputForm layout Expect BarcodeInputForm to have a Form with an input field and Button as child comps 1`] = `
<Form
  compact={false}
  fill={false}
  pad="none"
  plain={true}
>
  <style>
    
      #barcodeInputContainer .grommetux-button__icon {
        display: none;
      }

      #barcodeInputContainer .grommetux-number-input {
        display: block;
      }

      #barcodeInputContainer input {
        width: 100%;
      }

      input[type="number"]::-webkit-outer-spin-button,
      input[type="number"]::-webkit-inner-spin-button {
          -webkit-appearance: none;
          margin: 0;
      }
      input[type="number"] {
          -moz-appearance: textfield;
      }
    
  </style>
  <Box
    alignContent="stretch"
    announce={false}
    direction="row"
    focusable={true}
    pad="none"
    responsive={true}
    tag="div"
  >
    <div
      id="barcodeInputContainer"
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Enter barcode"
        value=""
      />
    </div>
    <Button
      fill={true}
      label="Search"
      method="push"
      onClick={[Function]}
      secondary={true}
      type="submit"
    />
  </Box>
</Form>
`;

exports[`Testing on BarcodeInputForm... Test on methods Expect componentDidMount and receiveFocus is called 1`] = `
<BarcodeInputForm
  onSubmitBarcode={[MockFunction]}
  shouldFocusField={false}
>
  <Form
    compact={false}
    fill={false}
    pad="none"
    plain={true}
  >
    <form
      className="grommetux-form grommetux-form--pad-none grommetux-form--plain"
    >
      <style>
        
      #barcodeInputContainer .grommetux-button__icon {
        display: none;
      }

      #barcodeInputContainer .grommetux-number-input {
        display: block;
      }

      #barcodeInputContainer input {
        width: 100%;
      }

      input[type="number"]::-webkit-outer-spin-button,
      input[type="number"]::-webkit-inner-spin-button {
          -webkit-appearance: none;
          margin: 0;
      }
      input[type="number"] {
          -moz-appearance: textfield;
      }
    
      </style>
      <Box
        alignContent="stretch"
        announce={false}
        direction="row"
        focusable={true}
        pad="none"
        responsive={true}
        tag="div"
      >
        <div
          className="grommetux-box grommetux-box--direction-row grommetux-box--align-content-stretch grommetux-box--responsive grommetux-box--pad-none"
          style={Object {}}
        >
          <div
            id="barcodeInputContainer"
          >
            <input
              autoFocus={true}
              onChange={[Function]}
              placeholder="Enter barcode"
              value=""
            />
          </div>
          <Button
            fill={true}
            label="Search"
            method="push"
            onClick={[Function]}
            secondary={true}
            type="submit"
          >
            <button
              className="grommetux-button grommetux-button--secondary grommetux-button--fill"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              type="submit"
            >
              <span
                className="grommetux-button__label"
              >
                Search
              </span>
            </button>
          </Button>
        </div>
      </Box>
    </form>
  </Form>
</BarcodeInputForm>
`;

exports[`Testing on BarcodeInputForm... Test on methods Expect componentWillReceiveProps is called when shouldFocusField change from false to true 1`] = `
<BarcodeInputForm
  onSubmitBarcode={[MockFunction]}
  shouldFocusField={true}
>
  <Form
    compact={false}
    fill={false}
    pad="none"
    plain={true}
  >
    <form
      className="grommetux-form grommetux-form--pad-none grommetux-form--plain"
    >
      <style>
        
      #barcodeInputContainer .grommetux-button__icon {
        display: none;
      }

      #barcodeInputContainer .grommetux-number-input {
        display: block;
      }

      #barcodeInputContainer input {
        width: 100%;
      }

      input[type="number"]::-webkit-outer-spin-button,
      input[type="number"]::-webkit-inner-spin-button {
          -webkit-appearance: none;
          margin: 0;
      }
      input[type="number"] {
          -moz-appearance: textfield;
      }
    
      </style>
      <Box
        alignContent="stretch"
        announce={false}
        direction="row"
        focusable={true}
        pad="none"
        responsive={true}
        tag="div"
      >
        <div
          className="grommetux-box grommetux-box--direction-row grommetux-box--align-content-stretch grommetux-box--responsive grommetux-box--pad-none"
          style={Object {}}
        >
          <div
            id="barcodeInputContainer"
          >
            <input
              autoFocus={true}
              onChange={[Function]}
              placeholder="Enter barcode"
              value=""
            />
          </div>
          <Button
            fill={true}
            label="Search"
            method="push"
            onClick={[Function]}
            secondary={true}
            type="submit"
          >
            <button
              className="grommetux-button grommetux-button--secondary grommetux-button--fill"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              type="submit"
            >
              <span
                className="grommetux-button__label"
              >
                Search
              </span>
            </button>
          </Button>
        </div>
      </Box>
    </form>
  </Form>
</BarcodeInputForm>
`;

exports[`Testing on BarcodeInputForm... Test on prop Expect onSubmitBarcode prop function is called when submit Button is clicked 1`] = `
<Form
  compact={false}
  fill={false}
  pad="none"
  plain={true}
>
  <style>
    
      #barcodeInputContainer .grommetux-button__icon {
        display: none;
      }

      #barcodeInputContainer .grommetux-number-input {
        display: block;
      }

      #barcodeInputContainer input {
        width: 100%;
      }

      input[type="number"]::-webkit-outer-spin-button,
      input[type="number"]::-webkit-inner-spin-button {
          -webkit-appearance: none;
          margin: 0;
      }
      input[type="number"] {
          -moz-appearance: textfield;
      }
    
  </style>
  <Box
    alignContent="stretch"
    announce={false}
    direction="row"
    focusable={true}
    pad="none"
    responsive={true}
    tag="div"
  >
    <div
      id="barcodeInputContainer"
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Enter barcode"
        value=""
      />
    </div>
    <Button
      fill={true}
      label="Search"
      method="push"
      onClick={[Function]}
      secondary={true}
      type="submit"
    />
  </Box>
</Form>
`;

exports[`Testing on BarcodeInputForm... Test on state Expect barcode state to change when input field is changed 1`] = `
<Form
  compact={false}
  fill={false}
  pad="none"
  plain={true}
>
  <style>
    
      #barcodeInputContainer .grommetux-button__icon {
        display: none;
      }

      #barcodeInputContainer .grommetux-number-input {
        display: block;
      }

      #barcodeInputContainer input {
        width: 100%;
      }

      input[type="number"]::-webkit-outer-spin-button,
      input[type="number"]::-webkit-inner-spin-button {
          -webkit-appearance: none;
          margin: 0;
      }
      input[type="number"] {
          -moz-appearance: textfield;
      }
    
  </style>
  <Box
    alignContent="stretch"
    announce={false}
    direction="row"
    focusable={true}
    pad="none"
    responsive={true}
    tag="div"
  >
    <div
      id="barcodeInputContainer"
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Enter barcode"
        value=""
      />
    </div>
    <Button
      fill={true}
      label="Search"
      method="push"
      onClick={[Function]}
      secondary={true}
      type="submit"
    />
  </Box>
</Form>
`;
